File: packages/client/test/cardinal-components.test.ts
/**
 * The generated client artifact.
 *
 * These assertions are about the *shape* of what the generator emits — sizes,
 * registry coverage, offset handling, endianness. The cross-language proof
 * lives in the golden vectors; this file catches a generator that produces
 * something self-consistent but wrong.
 */
import { describe, expect, it } from 'vitest';
import {
  CARDINAL_REGISTRY,
  Grabbable,
  Health,
  SCHEMA_HASH,
} from '../src/cardinal/components.generated.js';

describe('generated client components', () => {
  it('exposes an eight-hex-character schema hash', () => {
    expect(SCHEMA_HASH).toMatch(/^[0-9a-f]{8}$/);
  });

  it('registers every schema component by id', () => {
    expect([...CARDINAL_REGISTRY.keys()].sort((a, b) => a - b)).toEqual([1, 2]);
    expect(CARDINAL_REGISTRY.get(1)!.name).toBe('Health');
    expect(CARDINAL_REGISTRY.get(2)!.name).toBe('Grabbable');
  });

  it('reports constant byte sizes', () => {
    expect(CARDINAL_REGISTRY.get(1)!.bytes).toBe(8); // f32 + f32
    expect(CARDINAL_REGISTRY.get(2)!.bytes).toBe(16); // u32 + vec3
  });

  it('exports the elics component objects themselves', () => {
    expect(CARDINAL_REGISTRY.get(1)!.component).toBe(Health);
    expect(CARDINAL_REGISTRY.get(2)!.component).toBe(Grabbable);
  });

  it('describes its fields, so a flat value list can be rebuilt', () => {
    expect(CARDINAL_REGISTRY.get(2)!.fields).toEqual([
      { name: 'holderId', slots: 1 },
      { name: 'grabPoint', slots: 3 },
    ]);
  });

  it('round-trips Health through its codec', () => {
    const spec = CARDINAL_REGISTRY.get(1)!;
    const view = new DataView(new ArrayBuffer(spec.bytes));
    spec.encode(view, 0, { current: 12.5, max: 100 });
    expect(spec.decode(view, 0)).toEqual({ current: 12.5, max: 100 });
  });

  it('round-trips Grabbable, vec3 included', () => {
    const spec = CARDINAL_REGISTRY.get(2)!;
    const view = new DataView(new ArrayBuffer(spec.bytes));
    spec.encode(view, 0, { holderId: 42, grabPoint: [1, -2, 3.5] });
    const decoded = spec.decode(view, 0) as { holderId: number; grabPoint: number[] };
    expect(decoded.holderId).toBe(42);
    expect(decoded.grabPoint).toEqual([1, -2, 3.5]);
  });

  it('encodes at a non-zero offset without touching its neighbours', () => {
    // The batched frame packs records back to back, so every codec has to
    // honour the offset it is handed.
    const spec = CARDINAL_REGISTRY.get(1)!;
    const view = new DataView(new ArrayBuffer(spec.bytes + 8));
    view.setUint32(0, 0xdeadbeef, true);
    spec.encode(view, 4, { current: 1.5, max: 2.5 });
    expect(view.getUint32(0, true)).toBe(0xdeadbeef);
    expect(spec.decode(view, 4)).toEqual({ current: 1.5, max: 2.5 });
  });

  it('writes little-endian, matching the rest of the protocol', () => {
    const spec = CARDINAL_REGISTRY.get(2)!;
    const view = new DataView(new ArrayBuffer(spec.bytes));
    spec.encode(view, 0, { holderId: 1, grabPoint: [0, 0, 0] });
    expect(view.getUint8(0)).toBe(1); // low byte first
    expect(view.getUint8(3)).toBe(0);
  });
});

File: packages/server/test/cardinal_components_test.exs
defmodule IwsdkPhoenix.CardinalComponentsTest do
  use ExUnit.Case, async: true

  alias IwsdkPhoenix.Cardinal.{Grabbable, Health, Registry, CharacterGenome}

  describe "registry" do
    test "knows every schema component by id" do
      assert Enum.sort(Registry.ids()) == [1, 2, 3, 4]
      assert Registry.module_for(1) == Health
      assert Registry.module_for(2) == Grabbable
      assert Registry.module_for(3) == IwsdkPhoenix.Cardinal.Weather
      assert Registry.module_for(4) == CharacterGenome
    end

    test "returns nil for an unknown id rather than raising" do
      # An unknown id can only reach here as a bug — the join-time hash check
      # rules out schema drift — so it must degrade, not crash the room.
      assert Registry.module_for(9999) == nil
      assert Registry.byte_size_for(9999) == nil
    end

    test "reports constant byte sizes" do
      assert Registry.byte_size_for(1) == 8
      assert Registry.byte_size_for(2) == 16
      assert Registry.byte_size_for(4) == 13
    end

    test "exposes an eight-hex-character schema hash" do
      assert Registry.schema_hash() =~ ~r/^[0-9a-f]{8}$/
    end
  end

  describe "Health" do
    test "round-trips" do
      encoded = Health.encode(%Health{current: 12.5, max: 100.0})
      assert byte_size(encoded) == 8
      assert {:ok, %Health{current: 12.5, max: 100.0}} = Health.decode(encoded)
    end

    test "rejects a payload of the wrong size" do
      assert Health.decode(<<0, 0, 0>>) == :error
    end

    test "reports its declaration order, which struct key order does not preserve" do
      assert Health.field_order() == [:current, :max]
    end
  end

  describe "Grabbable" do
    test "round-trips, vec3 included" do
      value = %Grabbable{holder_id: 42, grab_point: %{x: 1.0, y: -2.0, z: 3.5}}
      assert {:ok, decoded} = value |> Grabbable.encode() |> Grabbable.decode()
      assert decoded.holder_id == 42
      assert decoded.grab_point == %{x: 1.0, y: -2.0, z: 3.5}
    end

    test "encodes little-endian" do
      encoded = Grabbable.encode(%Grabbable{holder_id: 1, grab_point: %{x: 0.0, y: 0.0, z: 0.0}})
      assert <<1, 0, 0, 0, _rest::binary>> = encoded
    end

    test "reports its declaration order" do
      assert Grabbable.field_order() == [:holder_id, :grab_point]
    end
  end
end
